### ISSUE TITLE
FR: Calibre plugin : sync-ing back Date Read & Rating

### ISSUE THREAD

**Issue by didierm** [koreader/koreader#6399]
I am not entirely sure what to make of https://github.com/koreader/koreader/pull/6177#issuecomment-645918265 , but a feature I use a lot in Calibre Companion (Android app : https://play.google.com/store/apps/details?id=com.multipie.calibreandroid), is sync'ing back the 'Read' and 'Date Read' information to Calibre.
For this purpose, CC allows you to define the Calibre 'Read' and 'Date Read' columns in its settings (in my case : '#read' and '#date_read').

As KR already logs this information (at the end of the document : [Mark as Read] and/or [Book Status]='Finished'), I was wondering whether the Calibre plugin would allow for an analogous setup.
Sync'ing back the book rating (end of document > [Book Status] > rating) would be so nice too.

**Comment by pazos** [comment]
I'm afraid I cannot help here. The main limitation I reached with the current implementation is the lack of usage/understanding of calibre. I just use calibre to strip DRM of the books I buy online and to remove some corporative BS. I don't keep any specific metadata beyond series & tags or use calibre as a book manager. Most of my documents are managed outside calibre.

I think it is doable and certainly can assist (a bit) on the smart device app interface, but have no interest in doing this. PRs are very welcome.

If somebody wants to contribute keep in mind:

- we load the whole JSON database in memory, so we strip the things we don't use, which are almost all of them (except the things used to navigate/search metadata)
- AFAICT the userdata field can grow a lot, it would be a good idea to test the implementation on a big library (1000-10000+ books) with a big userdata for each entry. I think it is not a good idea to store the whole userdata, just the bits we might want.

The most difficult part is figuring out a way to store the relevant data of fields that are, in most cases, mutable. Once some hardcore calibre user does that sending the info back to calibre shouldn't be a problem.

**Comment by NiLuJe** [comment]
The thread right below the comment you linked kind of show that there's approximately seven billion ways to have a "Read" column in Calibre, which, in addition to the technical quirks mentioned by @pazos, make this hard to get right. (You'd basically have to add a new setting, asking for the right custom column name, and quite possibly its type).

**Comment by NiLuJe** [comment]
Also: which way do you sync? When? For which books? What happens when there's a conflict?

e.g., I'm using the Kobo Utilities plugin to handle this, which does this both ways, and leaves the final decision to the user to handle conflict (or outright refuse a sync for specific book(s)). Which means I currently simply tag books manually in Nickel whenever I startup Nickel for USBMS. I might end up writing a small script that does that automatically by walking KOReader's history, and updating Nickel's DB accordingly.

(That's a small hint that this may be better/easier done in a Calibre plugin, although I have no idea how that would translate w/ the Smart Device Driver ;p)

**Comment by didierm** [comment]
@pazos 
[a] Due to my limited knowledge on how the Calibre plugin operates, I am hesitant to muddle the discussion. However, I do not have the impression that CC (Calibre Companion) loads all Calibre metadata. My usage model is, to send books to be read over the wireless connection to CC. CC has its own management system of grouping & sorting the books (based on the Calibre/books metadata). In this context, I am not sure why - for the purpose of sync'ing back data of individual books from the device to Calibre - it matters how large the Calibre database is (but I am dense).

CC : http://faq.calibrecompanion.co/index.php?action=artikel&cat=7&id=11&artlang=en
_"When CC connects as a wireless device, it asks calibre for any updated metadata for books already on the device. It does not ask calibre to send books not already on the device, primarily because it has no idea how much of your calibre library you want on your device."_

@NiLuJe 
[b] Settings : Indeed you need a new KR setting to link [Read] / [Date Read] (and [Rating]) values to Calibre columns, as I illustrated above. Types are already predefined to [boolean] & [date].
CC : http://faq.calibrecompanion.co/index.php?action=artikel&cat=7&id=40&artlang=en
_"Can I synchronize CC's Read and Date Read information with calibre?"_
[c] Sync'ing & conflicts : 
CC : http://faq.calibrecompanion.co/index.php?action=artikel&cat=7&id=43&artlang=en
_"What rules are used to decide whether CC's or calibre's value is used when syncing"_
[d] Sync'ing for which books : only for books on the device, see above [a].

Lots of information here : http://faq.calibrecompanion.co/index.php?action=show&cat=7


The reason why I submitted this FR is, because it is IMHO part of the natural flow of using Calibre as a content provider / single source of truth (why else use the Calibre plugin) : 
* add metadata in the library (you do not want to do this on the device) ;
* use the wireless connection to download books & their metadata to the device ;
* read books (you do not want to do this in Calibre) ;
* update Date Read after reading (part of your reading flow, you do not want to do this in Calibre ; automated in KR) ;
* rate books after reading (part of your reading flow, again you do not want to do this in Calibre) ;
* use the subsequent wireless connection (may be several weeks later), to automaticvally exchange/sync the updated metadata (e.g. Date Read, Rating) from the device to Calibre.

**Comment by pazos** [comment]
> [a] Due to my limited knowledge on how the Calibre plugin operates, I am hesitant to muddle the discussion.

You're very welcome to ask questions. It could help others to understand the code :smile: 

 > However, I do not have the impression that CC (Calibre Companion) loads all Calibre metadata.

I don't know how CC stores the metadata but I guess is probably in a SQLite DB. I was talking about the metadata that calibre sends when a book is added, that matches what calibre stores in the JSON file. As you see in https://github.com/koreader/koreader/blob/master/plugins/calibre.koplugin/metadata.lua#L115-L120 we receive all that metadata and discard most of them (listed in https://github.com/koreader/koreader/blob/master/plugins/calibre.koplugin/metadata.lua#L13-L35), including the user_metadata field that holds user custom columns.

> I am not sure why - for the purpose of sync'ing back data of individual books from the device to Calibre - it matters how large the Calibre database is.

When the device is connected calibre ask how many books are in the device and we send a response with a count. Then calibre ask the id for each device and we send https://github.com/koreader/koreader/blob/master/plugins/calibre.koplugin/metadata.lua#L142-L149

If calibre finds something that doesn't match its knowledge it asks for more metadata and we send it in https://github.com/koreader/koreader/blob/master/plugins/calibre.koplugin/metadata.lua#L152-L160

As you see we're manipulating an array of *n* books. Each book has a few fields. If the number of fields or their complexity grows up we're handling more KB/MB in memory. For handling a few key elements the current implementation is ok even for thousands of books (I downloaded a bunch of Project Gutenberg books to test) but if we want to store complex data in the book array we might want to switch to a relational database, which IMHO, is pain.

**Comment by NiLuJe** [comment]
> I might end up writing a small script that does that automatically by walking KOReader's history, and updating Nickel's DB accordingly.
> 

[Oh, hai!](http://trac.ak-team.com/trac/browser/niluje/Configs/trunk/Kindle/Kobo_Hacks/Patches/koreader_to_nickel.py)

**Comment by harmtemolder** [comment]
I explored the possibility of syncing read progress from KOReader to calibre through a [KOReader Sync Server](https://github.com/koreader/koreader-sync-server). These are my most important findings:

- The progress sync plugin identifies books by a partial MD5 hash of the actual file, calculated with [this `fastDigest` function](https://github.com/koreader/koreader/blob/8403154d4d36f4279828e0af90667a92685a4ebe/frontend/document/document.lua#L125) ([in case you prefer Python](https://git.sr.ht/~harmtemolder/koreader-calibre-plugin/tree/main/fast_digest.py))
- Pushing an EPUB from calibre to a device changes the contents of that EPUB (I believe parts of `metadata.opf` are merged into `content.opf`), so the hash of both files will differ
- Connecting the device to calibre after calibre made changes to the metadata (e.g. updating read progress from a KOReader Sync Server) will update the EPUB file on the device, therefore altering the hash again, breaking the connection to its previous entries on the server

Because of this, I have decided to abandon this attempt and instead will try my hand at syncing calibre directly with a connected device (much like the aforementioned Kobo Utilities plugin).

**Comment by pazos** [comment]
> I explored the possibility of syncing read progress from KOReader to calibre through a KOReader Sync Server

@harmtemolder :That's a bit offtopic for this plugin and probably can't be fixed without changes in calibre.

If you know which changes need to be sent to calibre you can do that within this plugin. You just need to:

1. When initializing a wireless inbox there's a routine that checks that books exist and prune the ones that were removed. For books that exist you can read its reading state, update the calibre metadata and that metadata will be sent back to calibre on sync.

2. If you want to do the same with USB transfers you can add a button somewhere that does the same without the need of wireless transfer. That is: select a dir with a calibre library, prune removed and update reading position and write that file into calibre json files. Calibre uses an UUID per book and these UUID's don't change even when the file changed and hashes differ.

Please give it a try if you wish. It should be way easier than other alternatives.

**Comment by Galunid** [comment]
> Connecting the device to calibre after calibre made changes to the metadata (e.g. updating read progress from a KOReader Sync Server) will update the EPUB file on the device, therefore altering the hash again, breaking the connection to its previous entries on the server

That's not entirely true. The md5 hash is calculated only the first time book is opened, and then it's stored in `bookname.sdr/metadata.lua`. As long as the md5 hash in that file doesn't change, the progress will sync, even if the file's hash changes.

**Comment by NiLuJe** [comment]
> 
> 
> > Connecting the device to calibre after calibre made changes to the metadata (e.g. updating read progress from a KOReader Sync Server) will update the EPUB file on the device, therefore altering the hash again, breaking the connection to its previous entries on the server
> 
> That's not entirely true. The md5 hash is calculated only the first time book is opened, and then it's stored in `bookname.sdr/metadata.lua`. As long as the md5 hash in that file doesn't change, the progress will sync, even if the file's hash changes.

I haven't actually checked, but if that's true, I'd arguably consider that a bug? ^^.

Don't we re-compute it if the mtime changed or something?

**Comment by Galunid** [comment]
> I haven't actually checked, but if that's true, I'd arguably consider that a (generally helpful ;p) bug? ^^.
>
> Don't we re-compute it if the mtime changed or something?

I believe it's actually intended, since you can highlight something in PDF and save it, the hash would change, and prevent further progress syncs. I vaguely recall comments mentioning it's intended. Not sure if there are some conditions to trigger recomputation. There are 2 fields holding md5 hash, so perhaps one of them changes, and the other one stays? But agreed, that's OT.

**Comment by harmtemolder** [comment]
OK, back on topic: I tried getting my hands on KOReader’s sidecar files from calibre over the wireless connection, but unfortunately ran into [this](https://github.com/koreader/koreader/blob/99045b431104904d3391a3d5ed01566356f76c24/plugins/calibre.koplugin/wireless.lua#L630):

```lua
function CalibreWireless:sendToCalibre(arg)
    logger.dbg("GET_BOOK_FILE_SEGMENT", arg)
    -- not implemented yet, we just send an invalid opcode to raise a control error in calibre.
    -- If we don't do this calibre will wait *a lot* for the file(s)
    self:sendJsonData('NOOP', {})
end
```

That means there are 2 things I can do:

1. Implement the `sendToCalibre` function so that KOReader’s calibre plugin can respond to my requests coming from calibre.
2. Follow @pazos’s suggestion and instead write the contents of KOReader’s sidecar files to `metadata.calibre`, so that they will be part of the usual sync.

I have no background in Lua or the KOReader codebase. How feasible do you think either option is?

**Comment by pazos** [comment]
sendToCalibre is intended to send books to calibre from your device library. I don't know what is your intention sending text files to calibre but `sendToCalibre` is not the solution. ie: metadata files are not part of the library that the wireless driver exposes to calibre, so they won't be displayed as available on the device and you cannot retreive them from calibre.

Exposing metadata to calibre should be done using readily available APIs. So 1 is a dead end. 2 is doable, but requires time and, specially, some knowledge of calibre metadata scheme.

> I have no background in Lua or the KOReader codebase. How feasible do you think either option is?

Feel free to play with the code in the calibre plugin. If you have any question this is probably the best place to ask :)

**Comment by Galunid** [comment]
@harmtemolder To add to what pazos said, https://github.com/koreader/koreader/blob/master/plugins/statistics.koplugin/main.lua is perhaps a good place to get some basic knowledge of lua and how to use SQLite

**Comment by harmtemolder** [comment]
> sendToCalibre is intended to send books to calibre from your device library. I don't know what is your intention sending text files to calibre but `sendToCalibre` is not the solution. ie: metadata files are not part of the library that the wireless driver exposes to calibre, so they won't be displayed as available on the device and you cannot retreive them from calibre.

I agree with you that `sendToCalibre` was not intended to send sidecar files, but I do not agree that it cannot be done: The [`get_file` function of the `SMART_DEVICE_APP` class](https://github.com/kovidgoyal/calibre/blob/587b113c84e06ccf1a1d5cf51609b523e7d65d29/src/calibre/devices/smart_device_app/driver.py#L1568) (which sends the `GET_BOOK_FILE_SEGMENT` request) just asks for a file by path and since sidecar files are named in a standardized fashion, it would not be too hard requesting them and then parsing them into calibre’s metadata from within calibre.

I’ll follow your suggestion to go with option 2 though, since that sounds more doable. Now trying to set up a development environment for KOReader on my Mac...

**Comment by pazos** [comment]
> I agree with you that sendToCalibre was not intended to send sidecar files, but I do not agree that it cannot be done: The get_file function of the SMART_DEVICE_APP class (which sends the GET_BOOK_FILE_SEGMENT request) just asks for a file by path and since sidecar files are named in a standardized fashion, it would not be too hard requesting them and then parsing them into calibre’s metadata from within calibre.

It is a **very, very, very** bad idea :)

Let me explain. The protocol is:

KOReader as a client <-> calibre as a server

When a client connects the server asks for a book count. The client needs to send a few strings for each book, including book's *logical path* (it's called lpath in our implementation). Then calibre could request more metadata for a given book if the entry isn't found on its db, lpath changed, etc.

The server uses the info from the client to populate the "Device" tab, which lists which books are available on device (they can also be available or the calibre library or not). Then the user, on the server, can request a given book from the client using `GET_BOOK_FILE_SEGMENT` opcode. It will use **lpath** to tell the client which book it wants (again lpath is reported by the client when it reports the book count)

So, to sum up: your proposal would require:

- add the relevant *.lua files to the list of files supported by the client
- iterate over a calibre library, searching for koreader metadata files
- for each metadata file generate a new entry in the books metadata table (with its own uuid, lpath and index)
- for each removed book you'll need to delete the entry that points to its metadata.

At the end, the user, on calibre, will find that the device with n books will contain up to n*2 entries (book + ko metadata for that book) and metadata files will be plain lua text files without any valid calibre metadata.

TL;DR: Nope. Albeit technically possible is not a good idea. calibre apis are wonderful and have its own metadata management. We shouldn't mix books with metadata.

**Comment by harmtemolder** [comment]
The way you propose is indeed a bad idea. The way I proposed would instead do something along these lines for each book in the book list, or maybe only the selected book, after pressing a sync button in calibre:

```python
connected_device = self.gui.device_manager.connected_device
with open('temp.epub.lua', mode='wb') as outfile:
    connected_device.get_file(connected_device._main_prefix + connected_device.books()[0].lpath.replace('.epub', '.sdr/metadata.epub.lua'), outfile)
```

I tested this on a USB device, and there it works as expected: `metadata.epub.lua` is downloaded from the device to the local directory. Without the need of adding anything to any library.

But as I said, I will try option 2 instead, as soon as I manage to set up an environment that lets me build KOReader.

**Comment by pazos** [comment]
@harmtemolder: Indeed I was talking about changes in the client, not the server, which I consider a black box. If you want to make changes in calibre, or a calibre plugin to support certain ko features that's very welcome, but certainly doesn't involve an integration on our end :)

> I tested this on a USB device, and there it works as expected: metadata.epub.lua is downloaded from the device to temp.epub.lua in the local directory. Without the need of adding anything to any library.

If the only reason to avoid this approach is the lack of support for wireless file uploads we can react to `GET_BOOK_FILE_SEGMENT` opcode and send books based on server args. It shouldn't be too hard, just something I didn't want to do because I usually pull content from calibre but never pushed anything.

**Comment by harmtemolder** [comment]
Oh wow, I only now realize that I completely misinterpreted this feature request’s title: it’s a request for a feature for the calibre plugin in KOReader and not a request for a KOReader plugin in calibre. Sorry about that, but we’re finally on the same page now.

I don’t think I’m the right person right now to work on the calibre plugin in KOReader. And that approach would probably mean a user needs to configure both KOReader and calibre (e.g. column mappings) for this to work. Approaching it from the calibre side does make more sense in my eyes:

I am willing to write a plugin for calibre (server) that reads metadata from a connected device’s sidecar files (client). In fact, for USB devices there is nothing stopping me from already doing this. And for wireless devices the only thing blocking this is the lack of support you mention, reacting to [`GET_BOOK_FILE_SEGMENT`](https://github.com/kovidgoyal/calibre/blob/83d9f0946bc2d402452629226ab0990aada36740/src/calibre/devices/smart_device_app/driver.py#L1568) with a byte stream of the requested `lpath`. Is that something you would be willing to add? And if so, how can I help?

**Comment by NiLuJe** [comment]
I definitely agree that this is way more viable on Calibre's side.

Which is essentially what I rely on (via the KoboUtilities plugin) with [this](https://github.com/koreader/koreader/issues/6399#issuecomment-681026114).

**Comment by pazos** [comment]
> Is that something you would be willing to add? And if so, how can I help?

Sure. I remember playing a bit with it and deciding it doesn't worth the time :p. Will look at it this weekend.

It should be a matter of something like:

```lua
function CalibreWireless:sendToCalibre(arg)
    local lfs = require("lfs")
    local inbox_dir = G_reader_settings:readSetting("inbox_dir")
    local path = inbox_dir .. "/" .. arg.lpath
    local file = io.open(path, "rb")
    local data = file:read("*all")
    file:close()
    local size = lfs.attributes(path, "size")
    -- SEND HERE:
    -- 1ST json data with opcode and options (including file size)
    -- 2ND binary data
end
```

I don't remember if we need to serialize the data, or close the socket after sending the file to restore the onReceive callback. Feel free to play with it if you wish.

Most of the current implementation is copied from https://github.com/shermp/UNCaGED/blob/master/uc/uc.go (albeit network things in lua are overcomplicated)

**Comment by harmtemolder** [comment]
I wrote a calibre plugin that syncs metadata from KOReader’s sidecars to calibre: https://git.sr.ht/~harmtemolder/koreader-calibre-plugin#koreader-calibre-plugin. Please let me know what you think.

It works for USB devices, so next step is to dive into the `CalibreWireless:sendToCalibre` function @pazos mentioned.
